LwM2M: Added new callback type for Executes: includes payload to support execute parameters

MaikVermeulen opened the pull request.

rlubos commented on include/net/lwm2m.h:
```diff
@@ -200,6 +199,24 @@ typedef int (*lwm2m_engine_set_data_cb_t)(uint16_t obj_inst_id,
+ * lwm2m_engine_register_exec_callback()
+ *
+ * @param[in] obj_inst_id Object instance ID generating the callback.
+ * @param[in] payload Pointer to data argument payload.
```
I'd add information here, that it can be NULL if no arguments were provided in the execute message.

rlubos commented on subsys/net/lib/lwm2m/ipso_pressure_sensor.c:
```diff
@@ -83,7 +83,7 @@ static void update_max_measured(uint16_t obj_inst_id, int index)
 }
 
-static int reset_min_max_measured_values_cb(uint16_t obj_inst_id)
+static int reset_min_max_measured_values_cb(uint16_t obj_inst_id, uint8_t *payload, uint16_t payload_len)
```
This line looks lengthy, please keep in mind 80 char/line limit.

rlubos replied:
Plus there are many more callbacks that would need an update, a quick grep for `INIT_OBJ_RES_EXECUTE` shows the following files:
* ipso_generic_sensor.c
* ipso_humidity_sensor.c
* lwm2m_obj_device.c
* lwm2m_obj_server.c

rlubos commented on subsys/net/lib/lwm2m/ipso_temp_sensor.c:
```diff
@@ -99,7 +99,7 @@ static void update_max_measured(uint16_t obj_inst_id, int index)
 }
 
-static int reset_min_max_measured_values_cb(uint16_t obj_inst_id)
+static int reset_min_max_measured_values_cb(uint16_t obj_inst_id, uint8_t *payload, uint16_t payload_len)
```
Same here. Not to repeat myself, please check compliance check result (https://github.com/zephyrproject-rtos/zephyr/pull/30551/checks?check_run_id=1523239427) which lists all of the issues.

rlubos commented on subsys/net/lib/lwm2m/ipso_timer.c:
```diff
@@ -287,8 +287,8 @@ static void timer_work_cb(struct k_work *work)
-static int timer_trigger_cb(uint16_t obj_inst_id)
-{
+static int timer_trigger_cb(uint16_t obj_inst_id, uint8_t *payload, uint16_t payload_len)
+{  
```
No trailing spaces, please.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -2839,6 +2839,8 @@ static int lwm2m_exec_handler(struct lwm2m_message *msg)
 	struct lwm2m_engine_obj_inst *obj_inst;
 	struct lwm2m_engine_res *res = NULL;
 	int ret;
+  uint8_t *payload;
```
Please use tab for indentation.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -2848,9 +2850,11 @@ static int lwm2m_exec_handler(struct lwm2m_message *msg)
 		return ret;
 	}
+  
+  payload = (uint8_t *) coap_packet_get_payload(msg->in.in_cpkt, &payload_len);
```
No space after cast:
```suggestion
  payload = (uint8_t *)coap_packet_get_payload(msg->in.in_cpkt, &payload_len);
```

rlubos commented on subsys/net/lib/lwm2m/lwm2m_obj_firmware.c:
```diff
@@ -298,7 +298,7 @@ static int firmware_update_cb(uint16_t obj_inst_id)
 	callback = lwm2m_firmware_get_update_cb();
 	if (callback) {
-		ret = callback(obj_inst_id);
+		ret = callback(obj_inst_id, NULL, 0);
```
You could just forward the `payload` and `payload_len` to the application callback.

rlubos commented on include/net/lwm2m.h:
```diff
@@ -200,6 +199,24 @@ typedef int (*lwm2m_engine_set_data_cb_t)(uint16_t obj_inst_id,
+ *         reason of failure or 0 for success.
+ */
+typedef int (*lwm2m_engine_execute_cb_t)(uint16_t obj_inst_id,
+                                         uint8_t *payload, uint16_t payload_len);
```
I wonder about the naming, given the message payload is only supposed to carry arguments in execute messages, perhaps it would be more verbose to call these parameters `args` and `args_len`.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_object.h:
```diff
@@ -336,7 +336,7 @@ struct lwm2m_engine_res {
 	lwm2m_engine_get_data_cb_t		pre_write_cb;
 	lwm2m_engine_set_data_cb_t		post_write_cb;
-	lwm2m_engine_user_cb_t			execute_cb;
+	lwm2m_engine_execute_cb_t			execute_cb;
```
Now `execute_cb` is indented too much.

rlubos reviewed (changes requested):
This could be a good enhancement, the PR needs some polishing though.

Apart of the changes listed in the review, you also need to update the following:
* The `lwm2m_client` sample registers a few exec callbacks, it needs an update.
* The LwM2M engine documentation needs an update as well: https://github.com/zephyrproject-rtos/zephyr/blame/master/doc/reference/networking/lwm2m.rst#L262
* Since it's an API change, we need an entry in the changelog: https://github.com/zephyrproject-rtos/zephyr/blob/master/doc/releases/release-notes-2.5.rst#api-changes

rlubos commented:
Also, please do not include the commits that are already on the master manually (https://github.com/zephyrproject-rtos/zephyr/pull/30551/commits/87cd480b1664d9cd77af9906c1e7795bcd54f05a), use rebase instead.

You commit message also needs a proper body and an signoff entry (plus the title is too long), here are the commit guidelines:
https://docs.zephyrproject.org/latest/contribute/index.html#commit-guidelines

MaikVermeulen commented:
@rlubos Thank you for the feedback. I indeed see I missed the commit guidelines and some of the documentation, probably due to my enthusiasm. I will get on it!

Commit:
net: lwm2m: Added execute arguments support

A dedicated LwM2M execute callback type has been implemented which
supports execute arguments. The lwm2m engine, lwm2m_client sample and
lwm2m objects have been updated accordingly. Also the API change has
been documented, and the lwm2m engine reference has been updated.

Fixes #30551.

Signed-off-by: Maik Vermeulen <maik.vermeulen@innotractor.com>

MaikVermeulen commented:
@rlubos I believe I have processed all your feedback. Could you check again?

rlubos commented on doc/reference/networking/lwm2m.rst:
```diff
@@ -254,14 +254,26 @@ To use the LwM2M library, start by creating an LwM2M client context
 .. code-block:: c
 
-	static int device_reboot_cb(uint16_t obj_inst_id)
+	static int resource_create_cb(uint16_t obj_inst_id)
```
This callback is not registered in this example code, I don't think there's need for it. Just please update the existing docs, we don't need to extend it in this PR.

rlubos commented on doc/reference/networking/lwm2m.rst:
```diff
@@ -254,14 +254,26 @@ To use the LwM2M library, start by creating an LwM2M client context
+
+.. code-block:: c
+
+	static int device_reboot_cb(uint16_t obj_inst_id,
+                              uint8_t *args, uint16_t args_len)
```
Let's keep the proper alignment in the sample code as well (align to the opening bracket).
```suggestion
	static int device_reboot_cb(uint16_t obj_inst_id, uint8_t *args, 
				    uint16_t args_len)
```

rlubos commented on doc/reference/networking/lwm2m.rst:
```diff
@@ -254,14 +254,26 @@ To use the LwM2M library, start by creating an LwM2M client context
+                              uint8_t *args, uint16_t args_len)
+	{
+		LOG_INF("Device rebooting. (Argument list (%u bytes): %s)",
+            args_len, log_strdup(args));
```
I don't think this will actually work, note that `args` is not a C-string (it's not NULL terminated). And `args` can be NULL pointer as well.

MaikVermeulen replied:
I will remove it. I expected it to be a C-string as the arguments are in ANBF format.

rlubos commented on include/net/lwm2m.h:
```diff
@@ -200,6 +199,25 @@ typedef int (*lwm2m_engine_set_data_cb_t)(uint16_t obj_inst_id,
+ *
+ * @param[in] obj_inst_id Object instance ID generating the callback.
+ * @param[in] args Pointer to execute arguments payload. (This can be
+ *            null if no arguments are provided)
```
```suggestion
 *            NULL if no arguments are provided).
```

rlubos commented on include/net/lwm2m.h:
```diff
@@ -200,6 +199,25 @@ typedef int (*lwm2m_engine_set_data_cb_t)(uint16_t obj_inst_id,
+ * @return Callback returns a negative error code (errno.h) indicating
+ *         reason of failure or 0 for success.
+ */
+typedef int (*lwm2m_engine_execute_cb_t)(uint16_t obj_inst_id,
+						uint8_t *args, uint16_t args_len);
```
Alignment issue (finetune with spaces if needed).

rlubos commented on samples/net/lwm2m_client/src/lwm2m-client.c:
```diff
@@ -144,7 +144,8 @@ static int init_led_device(void)
 }
 
-static int device_reboot_cb(uint16_t obj_inst_id)
+static int device_reboot_cb(uint16_t obj_inst_id,
+														uint8_t *args, uint16_t args_len)
```
There's definitely too many tabs in here. Zephyr uses 8 spaces/tab length. I suggest the following format here and in other places (note that it might look unaligned in GH diff preview due to github formatting, but it's actually properly aligned in the code):
```suggestion
static int device_reboot_cb(uint16_t obj_inst_id, uint8_t *args, 
			    uint16_t args_len)
```
![image](https://user-images.githubusercontent.com/22167799/101772016-1a6cde00-3aeb-11eb-8eca-55d4c3c49d65.png)

rlubos commented on subsys/net/lib/lwm2m/lwm2m_object.h:
```diff
@@ -336,7 +336,7 @@ struct lwm2m_engine_res {
 	lwm2m_engine_get_data_cb_t		pre_write_cb;
 	lwm2m_engine_set_data_cb_t		post_write_cb;
-	lwm2m_engine_user_cb_t			execute_cb;
+	lwm2m_engine_execute_cb_t     execute_cb;
```
`execute_cb` is still misaligned, please use tabs to align with other fields.
```suggestion
	lwm2m_engine_execute_cb_t		execute_cb;
```

rlubos reviewed (approved):
Thanks, the PR has really gotten into shape 👍 

Just a few more style issues (I guess you might need to adjust tab length in your editor).

rlubos reviewed (changes requested):
Sorry, just realized I approved early by accident, waiting for further updates.

rlubos commented on samples/net/lwm2m_client/src/lwm2m-client.c:
```diff
@@ -155,7 +156,8 @@ static int device_reboot_cb(uint16_t obj_inst_id)
 
-static int device_factory_default_cb(uint16_t obj_inst_id)
+static int device_factory_default_cb(uint16_t obj_inst_id,
+																			uint8_t *args, uint16_t args_len)
```
There's still an alignment issue here.

rlubos commented on samples/net/lwm2m_client/src/lwm2m-client.c:
```diff
@@ -167,7 +169,8 @@ static int device_factory_default_cb(uint16_t obj_inst_id)
 #if defined(CONFIG_LWM2M_FIRMWARE_UPDATE_PULL_SUPPORT)
-static int firmware_update_cb(uint16_t obj_inst_id)
+static int firmware_update_cb(uint16_t obj_inst_id,
+															uint8_t *args, uint16_t args_len)
```
Here.

rlubos commented on subsys/net/lib/lwm2m/ipso_generic_sensor.c:
```diff
@@ -98,7 +98,8 @@ static void update_max_measured(uint16_t obj_inst_id, int index)
 
-static int reset_min_max_measured_values_cb(uint16_t obj_inst_id)
+static int reset_min_max_measured_values_cb(uint16_t obj_inst_id,
+																						uint8_t *args, uint16_t args_len)
```
Here.

rlubos commented on subsys/net/lib/lwm2m/ipso_humidity_sensor.c:
```diff
@@ -83,7 +83,8 @@ static void update_max_measured(uint16_t obj_inst_id, int index)
 
-static int reset_min_max_measured_values_cb(uint16_t obj_inst_id)
+static int reset_min_max_measured_values_cb(uint16_t obj_inst_id,
+																						uint8_t *args, uint16_t args_len)
```
Here.

rlubos commented on subsys/net/lib/lwm2m/ipso_pressure_sensor.c:
```diff
@@ -83,7 +83,8 @@ static void update_max_measured(uint16_t obj_inst_id, int index)
 
-static int reset_min_max_measured_values_cb(uint16_t obj_inst_id)
+static int reset_min_max_measured_values_cb(uint16_t obj_inst_id,
+																						uint8_t *args, uint16_t args_len)
```
Here.

rlubos commented on subsys/net/lib/lwm2m/ipso_temp_sensor.c:
```diff
@@ -99,7 +99,8 @@ static void update_max_measured(uint16_t obj_inst_id, int index)
 
-static int reset_min_max_measured_values_cb(uint16_t obj_inst_id)
+static int reset_min_max_measured_values_cb(uint16_t obj_inst_id,
+																						uint8_t *args, uint16_t args_len)
```
Here.

rlubos commented on subsys/net/lib/lwm2m/ipso_timer.c:
```diff
@@ -287,7 +287,8 @@ static void timer_work_cb(struct k_work *work)
 
-static int timer_trigger_cb(uint16_t obj_inst_id)
+static int timer_trigger_cb(uint16_t obj_inst_id,
+														uint8_t *args, uint16_t args_len)
```
Here.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_obj_device.c:
```diff
@@ -128,7 +128,8 @@ static struct lwm2m_engine_res_inst *error_code_ri;
 
-static int reset_error_list_cb(uint16_t obj_inst_id)
+static int reset_error_list_cb(uint16_t obj_inst_id,
+																uint8_t *args, uint16_t args_len)
```
Here.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_obj_firmware.c:
```diff
@@ -272,19 +272,20 @@ lwm2m_engine_set_data_cb_t lwm2m_firmware_get_write_cb(void)
 
-static int firmware_update_cb(uint16_t obj_inst_id)
+static int firmware_update_cb(uint16_t obj_inst_id,
+															uint8_t *args, uint16_t args_len)
```
Here.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_obj_server.c:
```diff
@@ -97,7 +97,8 @@ static int disable_cb(uint16_t obj_inst_id)
 
-static int update_trigger_cb(uint16_t obj_inst_id)
+static int update_trigger_cb(uint16_t obj_inst_id,
+															uint8_t *args, uint16_t args_len)
```
And here.

rlubos reviewed (commented):
There're still some alignment issues, hopefully I've listed them all this time.

MaikVermeulen commented:
@rlubos Thank you for the reviews. I set the tab width to use actual tabs of length 8. What can I do in the future to better detect these problems myself? The automatic checks did not indicate problems anymore, or at least I thought?

rlubos reviewed (approved):
LGTM, thanks.

rlubos commented:
> @rlubos Thank you for the reviews. I set the tab width to use actual tabs of length 8. What can I do in the future to better detect these problems myself? The automatic checks did not indicate problems anymore, or at least I thought?

Issues like this (line lengths etc) should be listed as warnings in the compliance check results (https://github.com/zephyrproject-rtos/zephyr/runs/1531250676?check_suite_focus=true). To be honest I haven't even checked if the ones I posted were listed(they should be as they obviously broke the char/line limit).
